### ISSUE TITLE
What are the entries of a diagonal operator?

### ISSUE THREAD

**Issue by dlfivefifty** [JuliaApproximation/ContinuumArrays.jl#19]
I had been thinking of operators in the kernel sense, that is, 
$$
(K u)(x) = \int_a^b K(x,y) u(y) dy
$$
So that `K[x,y]` is equivalent to `K(x,y)`.

But then we are using diagonal operators. One might expect `QuasiDiagonal(one.(Inclusion(-1...1)))` to be the identity operator, and thats essentially how we've been using it. The issue is that the identity operator corresponds to `K(x,y) = δ(y-x)` so is actual infinite on the diagonal.

Is this the convention we want? E.g., `QuasiDiagonal(sqrt.(1-x.^2))[0.1,0.1] == ∞`?

**Comment by jagot** [comment]
I usually think of diagonal operators (mostly potentials in my case) as simply function multiplication:

```
g(x) = U(x) f(x)
```

where `U(x)` is the diagonal operator. If you want it inside an integral, then you'd need the Dirac delta function, which of course complicates things if you want to index the operator itself, but is that something you need often?

In CoulombIntegrals.jl, I calculate the repulsion between two electrons, `1/abs(r_1 - r_2)`, by solving Poisson's problem over-and-over again, each time the electrons' wave functions have changed. The solution is a _new_ potential in coefficient space, i.e. the distribution of e.g. electron 1 along `r_1` forms a potential for electron 2. This is also a function, but only known at discrete points, e.g. a diagonal matrix acting on the coefficients of electron 2. Should this still be a `QuasiDiagonal`? I don't think so, but at the same time, I usually want to merge all potentials into one before acting with the total operator to save FLOPS.

**Comment by dlfivefifty** [comment]
Right, that’s exactly how I’ve been thinking of them. There is no actual need for getindex except to make sure we are conceptualising them. 

Perhaps `getindex` is best thought of in terms of the delta function. that is `K[s,t]` is equivalent to $\langle \delta_s, K\delta_t \rangle$ or in other words:
$$
\langle \delta_s, \int K(x,y) \delta_t(y) dy \rangle = \langle \delta_s, K(x,t) \rangle = K(s,t)
$$

Perhaps the issue is the usage of `QuasiDiagonal` / `Diagonal`, as it hints that it puts the quasi-vector on the diagonal, where really we have `Inf * v` on the diagonal of `QuasiDiagonal(v)`. Is there another name for this operation?

**Comment by jagot** [comment]
Maybe something with `PointWise*`? In physics, we'd call this a _local_ operator, whereas non-diagonal kernels are _non-local_; I prefer this terminology to point-wise.

**Comment by dlfivefifty** [comment]
Yes these are local operators, but so are derivatives. `Pointwise` is too application specific. Perhaps follow ApproxFun and call it `Multiplication`?
